renku/core/commands/update.py

```diff
@@ -46,96 +51,192 @@ def update_workflows():
 
-    # TODO: Implement this properly with new database
-    # graph = Graph()
-    # outputs = graph.build(revision=revision, can_be_cwl=no_output, paths=paths)
-    # outputs = {node for node in outputs if graph.need_update(node)}
-    # if not outputs:
-    #     communication.echo("All files were generated from the latest inputs.")
-    #     sys.exit(0)
+    client = client_dispatcher.current_client
 
-    # # Check or extend siblings of outputs.
-    # outputs = siblings(graph, outputs)
-    # output_paths = {node.path for node in outputs if _safe_path(node.path)}
+    paths = paths or []
+    paths = get_relative_paths(base=client.path, paths=paths)
 
-    # # Get all clean nodes.
-    # input_paths = {node.path for node in graph.nodes} - output_paths
+    modified_activities = _get_modified_activities(client, activity_gateway)
+    ordered_activities = _get_ordered_downstream_activities(modified_activities, activity_gateway, paths)
 
-    # # Store the generated workflow used for updating paths.
-    # workflow = graph.as_workflow(input_paths=input_paths, output_paths=output_paths, outputs=outputs)
+    update_plan = _create_plan_from_activity_list(ordered_activities)
 
-    # execute_workflow(workflow, output_paths, command_name="update", update_commits=True)
+    execute_workflow(workflow=update_plan, command_name="update")
 
 
-@inject.autoparams()
-def execute_workflow(
-    workflow,
-    output_paths,
-    command_name,
-    update_commits,
-    client_dispatcher: IClientDispatcher,
-    activity_gateway: IActivityGateway,
+def _get_modified_activities(client, activity_gateway) -> Generator[Activity, None, None]:
+    """Return latest activities that one of their inputs is modified."""
+    latest_activities = activity_gateway.get_latest_activity_per_plan().values()
+
+    used_entities = (u.entity for a in latest_activities for u in a.usages)
+    modified, _ = get_modified_entities(entities=used_entities, repo=client.repo)
+    return (a for a in latest_activities if any(u.entity in modified for u in a.usages))
+
+
+def _get_ordered_downstream_activities(
+    starting_activities: Generator[Activity, None, None], activity_gateway: IActivityGateway, paths: List[str]
 ):
-    """Execute a Run with/without subprocesses."""
-    client = client_dispatcher.current_client
+    """Return an ordered list of activities so that an activities comes before all its downstream activities."""
+    graph = networkx.DiGraph()
 
-    wf, path = CWLConverter.convert(workflow, client.path)
-    # Don't compute paths if storage is disabled.
-    if client.check_external_storage():
-        # Make sure all inputs are pulled from a storage.
-        paths_ = (i.consumes.path for i in workflow.inputs)
-        client.pull_paths_from_storage(*paths_)
+    activities = set(starting_activities)
+    while activities:
+        activity = activities.pop()
+        child_activities = activity_gateway.get_downstream_activities(activity, max_depth=1)
```

Comment:
I think we can just get everything from the catalog directly instead of repeatedly querying it.